### PR TITLE
Send responses to original browser in correct order.

### PR DESCRIPTION
**Pull request by t4lz**
Set a `request_id` for each incoming request, the responses are only sent back to the browser once all earlier responses have been sent.

### CODE REVIEW

File: mirrord/agent/src/steal/connection.rs
Comment (meowjesty) [inline mirrord/agent/src/steal/connection.rs]: Bit of a weird pattern to not initialize this directly.

I would personally go for chaining operations, but could also move everything to inside the `if let`.

File: mirrord/agent/src/steal/connection.rs
Comment (eyalb181) [inline mirrord/agent/src/steal/connection.rs]:
```suggestion
    /// For a connection with pending requests, a priority queue with responses that were already
```

File: mirrord/agent/src/steal/connection.rs
Comment (eyalb181) [inline mirrord/agent/src/steal/connection.rs:694]: What's response.response?

File: mirrord/agent/src/steal/connection.rs
Comment (eyalb181) [inline mirrord/agent/src/steal/connection.rs]: Pretty weird name plus you're retrieving it from a hash called `http_write_streams`?

File: mirrord/agent/src/steal/connection.rs
Comment (eyalb181) [inline mirrord/agent/src/steal/connection.rs:706]: Is this supposed to be == and not <=?

File: mirrord/protocol/src/tcp.rs
Comment (eyalb181) [inline mirrord/protocol/src/tcp.rs]: Couldn't parse this sentence

File: mirrord/agent/src/steal/connection.rs
Comment (t4lz) [reply mirrord/agent/src/steal/connection.rs]: Can't `return` from the function from within closures, so I don't think we could use chained operation here.
How about this: https://github.com/t4lz/mirrord/pull/3/commits/4f631bf6b934e2162ccbcc769e24fbf40afc8b3d?

File: mirrord/agent/src/steal/connection.rs
Comment (t4lz) [reply mirrord/agent/src/steal/connection.rs:694]: It's the `InternalHttpResponse` of the `HttpResponse`. Should that member be renamed to `internal_response`?

File: mirrord/agent/src/steal/connection.rs
Comment (t4lz) [reply mirrord/agent/src/steal/connection.rs]: Is `http_tx` better? https://github.com/t4lz/mirrord/pull/3/commits/5642fda59668475f3ddca338ff5c35981e4b655c

File: mirrord/agent/src/steal/connection.rs
Comment (t4lz) [reply mirrord/agent/src/steal/connection.rs:706]: This is a good point, we should make sure it can't be `<`. `counter` holds the `request_id` of the first response not yet sent, so there should never be a response with a `request_id` lower than `counter`. If that holds (and we should make sure it does), then we'll only have `<=` when it's `==`, so I would write `==` so that it's clear that it cannot be `<`.
Added https://github.com/t4lz/mirrord/pull/3/commits/827c2bb753f494a2d9825456ba982ecaec15e728 to handle the case where a client (layer) sends the stealer a response with a duplicate `request_id` and make sure we do not insert a response with a ` request_id` lower than the corresponding counter to the corresponding queue.

File: mirrord/protocol/src/tcp.rs
Comment (t4lz) [reply mirrord/protocol/src/tcp.rs]: Yeah that's bad, how about:
```suggestion
/// A total order of responses in which the response with the LESSER `request_id` is the GREATER one.
```

File: mirrord/agent/src/steal/connection.rs
Comment (meowjesty) [reply mirrord/agent/src/steal/connection.rs]: Better, but this should be `match` instead of `if let` I think.

<- Not a fan of early returns like that though, may I try "closurefying" this function (we could also try it together)?

Comment (meowjesty) [reply mirrord/agent/src/steal/connection.rs]: Something like this:

```rust
match self
            .http_write_streams
            .get_mut(&response.connection_id)
            .and_then(|stream| {
                let counter = self
                    .http_request_counters
                    .entry(response.connection_id)
                    .or_insert(0);

                if response.request_id == *counter {
                    Some((
                        stream,
                        counter,
                        response
                            .response
                            .try_into()
                            .map_err(Self::handle_response_reconstruction_fail),
                    ))
                } else if response.request_id > *counter {
                    self.http_response_queues
                        .entry(response.connection_id)
                        .or_insert_with(|| BinaryHeap::with_capacity(8))
                        .push(response);

                    None
                } else {
                    error!(
                    "Got an http response with a request_id that was already received and sent to \
                    the browser for this connection. Dropping the duplicate response."
                    );

                    None
                }
            }) {
            Some((http_tx, counter, response)) => {
                let response = response?;
                http_tx
                    .write_all(&Self::response_to_bytes(response).await)
                    .await?; // TODO: handle error?

                *counter += 1;

                Ok(())
            }
            None => Ok(()),
        }
```

Comment (t4lz) [reply mirrord/agent/src/steal/connection.rs]: Ok, changing. Why is this better though?

Comment (meowjesty) [reply mirrord/agent/src/steal/connection.rs]: imo: Makes it easier to read code when it has a "flow", like every step does an operation, and we leave "breaks" to another place.

Like `A |> B |> C |> D` and if any step `|>` failed, then do `E`.

When you're dealing with `Option<T>` and stuff, the more you can work in the `Option` context the better, especially when using an expression based language like rust.

Comment (t4lz) [reply mirrord/agent/src/steal/connection.rs]: This code does not do the same thing as the original code :(
Rewriting the original code to be more functional.

File: mirrord/protocol/src/tcp.rs
Comment (eyalb181) [reply mirrord/protocol/src/tcp.rs]: I don't understand the word 'total' in this context, it's fine otherwise though

File: mirrord/agent/src/steal/connection.rs
Comment (eyalb181) [reply mirrord/agent/src/steal/connection.rs:706]: Oh okay. I misunderstood anyway, didn't realize it was the request_id, not the response id.

File: mirrord/agent/src/steal/connection.rs
Comment (eyalb181) [reply mirrord/agent/src/steal/connection.rs]: Yup

**Review by eyalb181 — APPROVED** [review]

File: mirrord/agent/src/steal/connection.rs
Comment (t4lz) [reply mirrord/agent/src/steal/connection.rs]: @eyalb181 since this is not a merge into main, and this is a comment about style, and we're all kind of blocked on merging this code, we decided to merge this PR like this, and @meowjesty will do a separate PR for giving this function more functional style before we merge `stealer-filter` into `main`, ok?

**Pull request merged**
